executor/src/subscriber.rs

```diff
@@ -120,144 +156,313 @@ impl Subscriber {
+        ret
     }
 
-    /// Downloads the payload from the worker and forwards the output to the
-    /// executor channel if the operation is successful. An error is returned
-    /// if we can't forward the output to the executor. If an irrecoverable error
-    /// has occurred while downloading the payload then this method panics.
-    #[instrument(level="debug", skip_all, fields(certificate_id = ?message.certificate.digest()), err)]
-    async fn download_payload_and_forward(&self, message: ConsensusOutput) -> SubscriberResult<()> {
-        // Fetch the certificate's payload from the workers. This is done via the
-        // block_waiter component. If the batches are not available in the workers then
-        // block_waiter will do its best to sync from the other peers. Once all batches
-        // are available, we forward the certificate to the Executor Core.
-        let result = Self::wait_on_payload(
-            self.metrics.clone(),
-            self.get_block_retry_policy.clone(),
-            self.store.clone(),
-            self.tx_get_block_commands.clone(),
-            message,
-        )
-        .await;
-
-        match result {
-            Ok(output) => {
-                if self.tx_executor.send(output).await.is_err() {
-                    return Err(SubscriberError::ClosedChannel(
-                        stringify!(self.tx_executor).to_owned(),
-                    ));
-                }
+    /// Fetches single payload from network
+    /// This future performs infinite retries and blocks until Batch is available
+    /// As an optimization it tries to download from local worker first, but then fans out
+    /// requests to remote worker if not found locally
+    #[instrument(level = "debug", skip_all, fields(digest = % digest, worker_id = % worker_id))]
+    async fn fetch_payload(
+        &self,
+        digest: BatchDigest,
+        worker_id: WorkerId,
+        workers: Vec<NetworkPublicKey>,
+    ) -> Batch {
+        if let Some(payload) = self.try_fetch_locally(digest, worker_id).await {
+            return payload;
+        }
+        let _timer = self.metrics.subscriber_remote_fetch_latency.start_timer();
+        let mut stagger = Duration::from_secs(0);
+        let mut futures = vec![];
+        for worker in workers {
+            let future = self.fetch_from_worker(stagger, worker, digest);
+            futures.push(future.boxed());
+            stagger += Duration::from_secs(1);
+        }
+        let (batch, _, _) = futures::future::select_all(futures).await;
+        batch
+    }
+
+    #[instrument(level = "debug", skip_all, fields(digest = % digest, worker_id = % worker_id))]
+    async fn try_fetch_locally(&self, digest: BatchDigest, worker_id: WorkerId) -> Option<Batch> {
+        let _timer = self.metrics.subscriber_local_fetch_latency.start_timer();
+        let worker = self.network.my_worker(&worker_id);
+        let payload = self.network.request_batch(digest, &worker).await;
```

Comment:
Out of curiosity, will this eventually timeout at some point if no response is received? Not sure how Anemo handles requests (maybe @bmwill can provide some pointers?)

Reply:
If the connection is broken then yes it'll eventually timeout, otherwise if you need a particular timeout for this rpc you'll want to wrap it in tokio::time::timeout().